### PR TITLE
Add a Dockerfile that builds the CLI tool

### PR DESCRIPTION
**Pull request by dcreager**
I think this should be set up to build the dependencies separately, so that Docker layer caching kicks in, skipping that step when they haven't changed.

cf #62

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,33 @@
+FROM haskell:8.6 as build
+WORKDIR /build
+RUN cabal new-update
+
+# Build our upstream dependencies after copying in only enough to tell cabal
+# what they are.  This will make these layers cache better even as we change the
+# code of semantic itself.
+COPY semantic.cabal /build/
```
Comment (tclem) [inline Dockerfile]: Since the `WORKDIR` was set on line 2, this and the other copies can be just:

`COPY semantic.cabal .`

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+# semantic itself.
+COPY . /build
+RUN cabal new-build semantic:exe:semantic
+
+# A fake `install` target until we can get `cabal new-install` to work
+RUN cp $(find dist-newstyle -name semantic -type f -perm -u=x) /usr/local/bin/semantic
+
+# Create a fresh image containing only the compiled CLI program, so that the
+# image isn't bulked up by all of the extra build state.
+FROM debian:stretch
```
Comment (tclem) [inline Dockerfile]: Might do `debian:stable-slim` or `debian:stretch-slim` here.

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+RUN apt-get update && \
+  apt-get install -y \
+    libgmp10 \
+    && \
+  apt-get autoremove -y && \
+  apt-get clean -y && \
+  rm -rf /var/lib/apt/lists/*
+COPY --from=build /usr/local/bin/semantic /usr/local/bin/semantic
+ENTRYPOINT ["/usr/local/bin/semantic"]
```
Comment (tclem) [inline Dockerfile]: It would be good style to have some new lines between the directives here, I wonder if we've got a good dockerfile linter...

Comment (dcreager) [reply Dockerfile]: Done

File: Dockerfile (hunk 1 above)
Comment (dcreager) [reply Dockerfile]: Done

File: Dockerfile (hunk 2 above)
Comment (dcreager) [reply Dockerfile]: Done.  (Used `stretch-slim` instead of `stable-slim` since `haskell:8.6` is based off of `stretch`.)

**Pull request merged**